File: bin/wideSCCluster2longSCCluster.R
#!/usr/bin/env Rscript

library(optparse)
library(tidyr)
library(data.table)

option_list <- list( 
  make_option(c("-c", "--clusters-file"), dest="clusters_path"),
  make_option(c("-e", "--experiment-accession"), dest="exp_acc"),
  make_option(c("-o", "--output"), dest="output_path")
)

opt <- parse_args(OptionParser(option_list=option_list))

fread(opt$clusters_path, header = TRUE, check.names = FALSE, sep = '\t') -> clusters_wide
gather(clusters_wide, key = "cell_id", value = "cluster_id", -sel.K, -K) -> clusters_long

clusters_long$experiment_accession <- opt$exp_acc
names(clusters_long)[names(clusters_long) == 'K'] <- 'k'

fwrite(clusters_long[,c('experiment_accession','cell_id','k','cluster_id')],
          file=opt$output_path,
          row.names = FALSE,
          quote = TRUE)

File: bin/wideSCCluster2longSCCluster.py
#!/usr/bin/env python

import argparse
import pandas as pd
import csv

# Parse command-line arguments
parser = argparse.ArgumentParser()
parser.add_argument("-c", "--clusters-file", dest="clusters_path", help="Path to clusters file")
parser.add_argument("-e", "--experiment-accession", dest="exp_acc", help="Experiment accession")
parser.add_argument("-o", "--output", dest="output_path", help="Output file path")
args = parser.parse_args()

# Read clusters file
clusters_wide = pd.read_csv(args.clusters_path, sep='\t', header=0, usecols=lambda x: x != "sel.K")

# Reshape data from wide to long format
clusters_long = pd.melt(clusters_wide, id_vars=['K'], var_name='cell_id', value_name='cluster_id')

# Add experiment accession column
clusters_long['experiment_accession'] = args.exp_acc

# Rename 'K' column to 'k'
clusters_long.rename(columns={'K': 'k'}, inplace=True)

# Select and reorder columns
columns = ['experiment_accession', 'cell_id', 'k', 'cluster_id']

# Write output to file
clusters_long.to_csv(args.output_path, index=False, columns=columns, quoting=csv.QUOTE_NONNUMERIC)
